### PR TITLE
lint: switch black+import+flake8 -> Ruff

### LINKED ISSUE #253: unify various linters with Ruff
Seems that the pre-commit is calling several linters but they can be replaced with Ruff:
- `flake8` (native with codes `F`, `E`)
- imports (with codes `I`)
- `black` with `ruff-format`

Happy to send PR if you are fine with this change, @hugovk

### PR DESCRIPTION
**Pull request by Borda**
resolves #253

cc: @hugovk

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,3 +3,45 @@ requires = [
...
+#    "A",    # see: https://pypi.org/project/flake8-builtins
+#    "B",    # see: https://pypi.org/project/flake8-bugbear
+#    "C4",   # see: https://pypi.org/project/flake8-comprehensions
+#    "PT",   # see: https://pypi.org/project/flake8-pytest-style
+#    "RET",  # see: https://pypi.org/project/flake8-return
+#    "ANN",  # see: https://pypi.org/project/flake8-annotations
+#]
+#lint.ignore = [
+#    "E731",   # Do not assign a lambda expression, use a def
+#]
```
Comment (hugovk) [inline pyproject.toml:31]: Remove unused
```suggestion
```

File: pyproject.toml [hunk 2]
```diff
@@ -3,3 +3,45 @@ requires = [
...
+lint.ignore-init-module-imports = true
+lint.unfixable = ["F401"]
```
Comment (hugovk) [inline pyproject.toml:33]: Don't seem to be needed?
```suggestion
```

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -15,17 +10,13 @@ repos:
...
-      rev: v3.12.0
+    - repo: https://github.com/astral-sh/ruff-pre-commit
+      rev: v0.2.0
       hooks:
-          - id: reorder-python-imports
-            args: ['--application-directories=.:src', --py3-plus]
+        - id: ruff-format
+          args: ["--preview"]
+        - id: ruff
+          args: ["--fix"]
```
Comment (hugovk) [inline .pre-commit-config.yaml:19]: Move this...
```suggestion
```

File: pyproject.toml [hunk 4]
```diff
@@ -3,3 +3,45 @@ requires = [
     "setuptools >= 40.0",
 ]
 build-backend = "setuptools.build_meta"
+
+
+
+[tool.ruff]
```
Comment (hugovk) [inline pyproject.toml:9]: ... to here?

```suggestion
[tool.ruff]
fix = true
```

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -15,17 +10,13 @@ repos:
...
-    - repo: https://github.com/PyCQA/flake8
-      rev: 7.0.0
-      hooks:
-          - id: flake8
-            language_version: python3
-            additional_dependencies: [flake8-typing-imports>=1.9.0]
-    - repo: https://github.com/asottile/reorder_python_imports
-      rev: v3.12.0
+    - repo: https://github.com/astral-sh/ruff-pre-commit
+      rev: v0.2.0
```
Comment (hugovk) [inline .pre-commit-config.yaml:14]:
```suggestion
      rev: v0.2.2
```

File: pyproject.toml [hunk 6]
```diff
@@ -3,3 +3,45 @@ requires = [
...
+target-version = "py38"
```
[the remark this answers is not in the document]
Comment (Borda) [reply pyproject.toml:10]:
```suggestion
# todo: drop when migrated project metadata from setup.cfg
target-version = "py38"
```

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -15,17 +10,12 @@ repos:
...
-      rev: v3.12.0
+    - repo: https://github.com/astral-sh/ruff-pre-commit
+      rev: v0.2.2
       hooks:
-          - id: reorder-python-imports
-            args: ['--application-directories=.:src', --py3-plus]
+        - id: ruff-format
+          args: ["--preview"]
+        - id: ruff
     - repo: https://github.com/asottile/pyupgrade
```
[the remark this answers is not in the document]
Comment (Borda) [reply .pre-commit-config.yaml]: done :)

**Comment by Borda** [comment]
it just needs #254

**Comment by hugovk** [comment]
#254 has been merged, please could you resolve conflicts?

**Comment by Borda** [comment]
> please could you resolve conflicts?

@hugovk done :rabbit:

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -15,22 +10,12 @@ repos:
...
-    - repo: https://github.com/asottile/pyupgrade
-      rev: v3.15.1
+    - repo: https://github.com/astral-sh/ruff-pre-commit
+      rev: v0.2.2
       hooks:
-          - id: pyupgrade
-            args: [--py38-plus]
+        - id: ruff-format
+          args: ["--preview"]
+        - id: ruff
```
Comment (hugovk) [inline .pre-commit-config.yaml:18]: Let's follow the lead of https://github.com/sphinx-contrib/sphinx-lint/pull/110 and add:

```suggestion
        - id: ruff
          args: ["--exit-non-zero-on-fix"]
```

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -15,22 +10,13 @@ repos:
...
-            language_version: python3
-            additional_dependencies: [flake8-typing-imports>=1.9.0]
-    - repo: https://github.com/asottile/reorder_python_imports
-      rev: v3.12.0
-      hooks:
-          - id: reorder-python-imports
-            args: ['--application-directories=.:src', --py3-plus]
-    - repo: https://github.com/asottile/pyupgrade
-      rev: v3.15.1
+    - repo: https://github.com/astral-sh/ruff-pre-commit
```
Comment (hugovk) [inline .pre-commit-config.yaml:13]: Hopefully the final suggestion :)

Let's put the Ruff repo before the pre-commit-hooks one.

Reasoning: Ruff might do some formatting/fixes that prevent pre-commit-hooks failing.

**Review by hugovk — APPROVED** [review]
Thank you!

**Pull request merged**